daemon/cluster/cluster.go

```diff
@@ -893,6 +893,17 @@ func (c *Cluster) UpdateService(serviceID string, version uint64, spec types.Ser
 		return err
 	}
 
+	// Check if anything is modified in spec for update #25770
+	currentService, err := getService(ctx, c.client, serviceID)
+	if err != nil {
+		return err
+	}
+
+	if serviceSpec.String() == currentService.String() {
```

Comment:
`serviceSpec` is `swarmapi.ServiceSpec`:
https://github.com/kunalkushwaha/docker/blob/ad6a1e09b2f7d2af01006e71c8f5542c71fffdcb/daemon/cluster/cluster.go#L834
https://github.com/kunalkushwaha/docker/blob/ad6a1e09b2f7d2af01006e71c8f5542c71fffdcb/daemon/cluster/convert/service.go#L96

However, `currentService` is `*swarmapi.Service`: https://github.com/kunalkushwaha/docker/blob/ad6a1e09b2f7d2af01006e71c8f5542c71fffdcb/daemon/cluster/helpers.go#L53

So I wonder if this comparison is robust.

Also, I fear the comparison of `String()` can fail intermittently if the structure contains map in future versions.